src/node/stage/approved.rs

```diff
@@ -300,115 +296,273 @@ impl Approved {
-                    return Ok(MessageAction::Discard);
-                }
-
-                match &accumulating_msg.content.variant {
-                    Variant::NeighbourInfo { .. }
-                    | Variant::UserMessage(_)
-                    | Variant::NodeApproval(_)
-                    | Variant::Relocate(_) => Ok(MessageAction::Handle),
-
-                    Variant::GenesisUpdate(info) => {
-                        if self.should_handle_genesis_update(our_id, info) {
-                            Ok(MessageAction::Handle)
-                        } else if is_self_elder {
-                            Ok(MessageAction::Bounce)
-                        } else {
-                            Ok(MessageAction::Discard)
-                        }
-                    }
-
-                    // These variants are not to be signature-accumulated
-                    Variant::MessageSignature(_)
-                    | Variant::BootstrapRequest(_)
-                    | Variant::BootstrapResponse(_)
-                    | Variant::JoinRequest(_)
-                    | Variant::MemberKnowledge(_)
-                    | Variant::ParsecRequest(..)
-                    | Variant::ParsecResponse(..)
-                    | Variant::Ping
-                    | Variant::Bounce { .. } => Ok(MessageAction::Discard),
+            Variant::MessageSignature(_) => {
+                if self.verify_message(msg)? {
+                    Ok(MessageStatus::Useful)
+                } else {
+                    Ok(MessageStatus::Useless)
                 }
             }
             Variant::JoinRequest(req) => {
                 if !self.should_handle_join_request(our_id, req) {
-                    return Ok(MessageAction::Bounce);
+                    // Note: We don't bounce this message because the current bounce-resend
+                    // mechanism wouldn't preserve the original SocketAddr which is needed for
+                    // properly handling this message.
+                    // This is OK because in the worst case the join request just timeouts and the
+                    // joining node sends it again.
+                    return Ok(MessageStatus::Useless);
                 }
 
-                if !self.verify_message(msg)? {
-                    return Ok(MessageAction::Discard);
+                if self.verify_message(msg)? {
+                    Ok(MessageStatus::Useful)
+                } else {
+                    Ok(MessageStatus::Useless)
                 }
-
-                Ok(MessageAction::Handle)
             }
             Variant::BootstrapRequest(_)
-            | Variant::MemberKnowledge(_)
+            | Variant::ParsecPoke(_)
             | Variant::ParsecRequest(..)
             | Variant::ParsecResponse(..)
-            | Variant::Bounce { .. } => {
+            | Variant::BouncedUntrustedMessage(_)
+            | Variant::BouncedUnknownMessage { .. } => {
                 if self.verify_message(msg)? {
```

Comment:
Maybe, but not sure it would end up being cleaner. This way at least you can immediatelly see how every message variant is handled. I can change it if you feel strongly about it.

Reply:
well, that's the same block of code being used for 3-4 times.
regarding `immediatelly see how every message variant is handled`, i think that will still be in each match branch?